datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -781,10 +757,11 @@ pub async fn from_substrait_rel(
                     from_substrait_named_struct(named_struct, extensions)?
                         .replace_qualifier(table_reference.clone());
 
+                let substrait_schema =
+                    apply_projection(substrait_schema, &read.projection)?;
```

Comment:
Looking to future Substrait compatabilty. Substrait ReadRels have a [filter](https://github.com/substrait-io/substrait/blob/a0388ff69a83aa4addf51236b77bf275b3647590/proto/substrait/algebra.proto#L69) field that represents filter expressions that are applied _after_ the read but _before_ the masking (i.e. the filter expression can reference fields that are later masked out). 

Is the solution you're implementing here going to be compatible with that? By applying the masking here, we might potentially mask out fields that we're going to need.

Reply:
that's a good point. TableScan in DF seems to have a `filters` field, which I'm assuming is supposed to work the same way as substrait's filter, so I don't think we're in danger of doing that. Although this probably means that even the part of the base_schema that is being masked out needs some sort of validation because ReadRel's filter (and/or best_effort_filter) field might operate on them. This probably decides the question I raised above whether plans with inconsistent data types in masked out fields should be accepted or not.